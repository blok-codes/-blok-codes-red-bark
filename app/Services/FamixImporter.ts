import { inject, injectable } from 'inversify';
import { Project } from 'ts-morph';
import { Logger } from 'winston';

@injectable()
export class FamixImporter {
    @inject('Logger')
    private readonly logger!: Logger;

    @inject('Project')
    private readonly project!: Project;

    public readonly import = (path: string): void => {
        this.logger.info(`Importing project at ${path}`);
        this.project.addSourceFilesAtPaths(`${path}/**/*.ts`);
    };

    public readonly getJsonModel = (): string => '[]';
}
