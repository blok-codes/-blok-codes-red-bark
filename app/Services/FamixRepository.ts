import { FamixBaseElement } from '@blok-codes/famix/dist/FamixBaseElement';
import { FamixRepository as AbstractFamixRepository } from '@blok-codes/famix/dist/FamixRepository';
import { Class } from '@blok-codes/famix/dist/FamixTypeScript/Class';
import { Namespace } from '@blok-codes/famix/dist/FamixTypeScript/Namespace';
import { decorate, injectable } from 'inversify';

decorate(injectable(), AbstractFamixRepository);

@injectable()
export class FamixRepository extends AbstractFamixRepository {
    private static instance: FamixRepository;
    protected counter = 0;

    protected readonly records = {
        classes: new Set<Class>(),
        elements: new Set<FamixBaseElement>(),
        namespaces: new Set<Namespace>(),
    };

    public static readonly getInstance = (): FamixRepository => {
        if (!FamixRepository.instance) {
            FamixRepository.instance = new FamixRepository();
        }

        return FamixRepository.instance;
    };

    public readonly addElement = (element: FamixBaseElement): void => {
        if (element instanceof Class) {
            this.records.classes.add(element);
        } else if (element instanceof Namespace) {
            this.records.namespaces.add(element);
        } else {
            this.records.elements.add(element);
        }

        element.id = ++this.counter;
    };

    public static readonly clear = (): void => {
        FamixRepository.instance = new FamixRepository();
    };

    public readonly createOrGetFamixClass = (name: string, isAbstract = false): Class => {
        let instance = this.getFamixClass(name);

        if (!instance) {
            instance = new Class(this);

            instance.name = name;
            instance.isStub = true;

            instance.isAbstract = isAbstract;
        }

        return instance;
    };

    public readonly getFamixClass = (name: string): Class | undefined =>
        Array.from(this.records.classes.values()).find((c) => c.name === name);

    public readonly getFamixElementById = (id: number): FamixBaseElement | undefined =>
        Array.from(this.records.elements.values()).find((e) => e.id === id);

    public readonly getFamixElementsByType = (type: string): Set<FamixBaseElement> =>
        new Set(Array.from(this.records.elements.values()).filter((e) => e.constructor.name === type));

    public readonly getFamixFunction = (namespace: string, regex: string): FamixBaseElement | undefined =>
        Array.from(this.records.elements).find(
            // eslint-disable-next-line @typescript-eslint/no-explicit-any
            (e) => e instanceof Function && e.name.match(regex) && (e as any).namespace
        );

    public readonly getFamixNamespace = (name: string): Namespace | undefined =>
        Array.from(this.records.namespaces.values()).find((n) => n.name === name);

    public readonly toJSON = (): string => {
        let buffer = '[';

        for (const clazz of this.records.classes) {
            buffer = buffer + clazz.toJSON() + ',';
        }

        for (const element of this.records.elements) {
            buffer = buffer + element.toJSON() + ',';
        }

        buffer = buffer.substring(0, buffer.length - 1);
        return buffer + ']';
    };
}
