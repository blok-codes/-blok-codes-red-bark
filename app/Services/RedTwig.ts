import { FamixRepository } from '@blok-codes/famix/dist/FamixRepository';
import { Class } from '@blok-codes/famix/dist/FamixTypeScript/Class';
import { Method } from '@blok-codes/famix/dist/FamixTypeScript/Method';
import { inject, injectable } from 'inversify';
import {
    ClassDeclaration,
    FunctionDeclaration,
    MethodDeclaration,
    ModuleDeclaration,
    Project,
    SourceFile,
    Statement,
    SyntaxKind,
    VariableDeclaration,
    VariableStatement,
} from 'ts-morph';
import { Logger } from 'winston';

import { isProcessableStatement } from '../Helpers';

@injectable()
export class RedTwig {
    @inject('Logger')
    private readonly logger!: Logger;

    @inject('Project')
    private readonly project!: Project;

    @inject('FamixRepository')
    private readonly repository!: FamixRepository;

    // todo: delete these fields - placeholders until famix elements are created and stored in the repository
    private readonly classes: ClassDeclaration[] = [];
    private readonly declarations: VariableDeclaration[] = [];

    private readonly functions: FunctionDeclaration[] = [];
    private readonly methods: MethodDeclaration[] = [];

    private readonly variables: VariableDeclaration[] = [];
    private readonly variableStatements: VariableStatement[] = [];

    public readonly import = (paths: string[]): void => {
        this.logger.info(`Importing project at `, paths);
        this.repository.toJSON(); // todo: remove this line - just to compile

        const files = this.project.addSourceFilesAtPaths(paths);
        files.forEach((sourceFile: SourceFile) => this.process(sourceFile));

        this.report(); // todo: delete this line - just to debug for now
    };

    // todo: delete this method - just to debug for now
    private readonly report = (): void => {
        this.logger.info(`\nClasses:`);
        this.classes.forEach((c) => this.logger.debug(c.getName()));

        this.logger.info(`\nVariables:`);
        this.variables.forEach((v) => this.logger.debug(v.getName()));

        this.logger.info(`\nVariable Statements:`);
        this.variableStatements.forEach((vs) => this.logger.debug(vs.getDeclarations()[0].getName()));

        this.logger.info(`\nFunctions:`);
        this.functions.forEach((f) => this.logger.debug(f.getName()));

        this.logger.info(`\nMethods:`);
        this.methods.forEach((f) => this.logger.debug(f.getName()));

        this.logger.info(`\nVariable Declarations:`);
        this.declarations.forEach((v) => this.logger.debug(v.getName()));
    };

    private readonly process = (sourceFile: SourceFile): void => {
        this.logger.info(`----------Finding Classes:`);
        sourceFile.getClasses().forEach((declaration) => this.processClassDeclaration(declaration));

        this.logger.info(`----------Finding VariableStatements:`);
        sourceFile.getVariableStatements().forEach((statement) => this.processVariableStatement(statement));

        this.logger.info(`----------Finding VariableDeclarations:`);
        sourceFile.getVariableDeclarations().forEach((declaration) => this.processVariableDeclaration(declaration));

        this.logger.info(`----------Finding Functions:`);
        sourceFile.getFunctions().forEach((declaration) => this.processFunctionDeclaration(declaration));

        this.logger.info(`----------Finding Modules:`);
        sourceFile.getModules().forEach((declaration) => this.processModuleDeclaration(declaration));
    };

    private readonly processClassDeclaration = (declaration: ClassDeclaration): void => {
        this.classes.push(declaration);
        const clazz = this.repository.createOrGetFamixClass(declaration.getName());

        this.repository.addElement(clazz);
        declaration.getMethods().forEach((method) => this.processMethodDeclaration(method, clazz));

        this.logger.debug(
            `class: ${declaration.getName()} (${declaration.getType().getText()}), fqn = ${declaration
                .getSymbol()
                ?.getFullyQualifiedName()}`
        );
    };

    private readonly processMethodDeclaration = (declaration: MethodDeclaration, clazz: Class): void => {
        this.methods.push(declaration);
        this.logger.debug(`method: ${declaration.getName()}`);

        const method = new Method(this.repository);
        method.isAbstract = declaration.isAbstract();

        clazz.addMethods(method);
    };

    private readonly processFunctionDeclaration = (declaration: FunctionDeclaration): void => {
        this.functions.push(declaration);
        this.logger.debug(`function: ${declaration.getName()}`);

        declaration.getVariableStatements().forEach((statement) => this.processVariableStatement(statement));
        declaration.getVariableDeclarations().forEach((variable) => this.processVariableDeclaration(variable));

        this.processStatements(declaration.getStatements());
        declaration.getFunctions().forEach((func) => this.processFunctionDeclaration(func));
    };

    private readonly processModuleDeclaration = (declaration: ModuleDeclaration): void => {
        this.logger.debug(`Module "${declaration.getName()}":`);

        this.logger.info(`----------Finding Classes in Module "${declaration.getName()}":`);
        declaration.getClasses().forEach((clazz) => this.processClassDeclaration(clazz));

        this.logger.info(`----------Finding VariableStatements in Module "${declaration.getName()}":`);
        declaration.getVariableStatements().forEach((statement) => this.processVariableStatement(statement));

        this.logger.info(`----------Finding VariableDeclarations in Module "${declaration.getName()}":`);
        declaration.getVariableDeclarations().forEach((variable) => this.processVariableDeclaration(variable));

        this.logger.info(`----------Finding Functions in Module "${declaration.getName()}":`);
        declaration.getFunctions().forEach((func) => this.processFunctionDeclaration(func));
    };

    private readonly processVariableDeclaration = (declaration: VariableDeclaration): void => {
        this.variables.push(declaration);
        this.logger.debug(
            `variable: ${declaration.getName()} (${declaration.getType().getText()}), fqn = ${declaration
                .getSymbol()
                ?.getFullyQualifiedName()}`
        );
    };

    private readonly processVariableStatement = (statement: VariableStatement): void => {
        this.variableStatements.push(statement);
        this.logger.debug(
            `Variable statement: ${statement.getDeclarationKindKeyword().getText()} ${statement
                .getDeclarations()[0]
                .getName()}`
        );
    };

    private readonly processStatements = (statements: Statement[]): void => {
        this.logger.debug(`processing statements...`);

        statements
            .filter((statement) => isProcessableStatement(statement))
            .forEach((statement) => {
                this.logger.debug(`variables in ${statement.getKindName()}`);

                statement
                    .getDescendantsOfKind(SyntaxKind.VariableDeclaration)
                    .forEach((declaration: VariableDeclaration) => {
                        this.processVariableDeclaration(declaration);
                        this.declarations.push(declaration);
                    });
            });
    };
}
