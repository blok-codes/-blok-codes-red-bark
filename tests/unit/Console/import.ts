import { FamixRepository } from '@blok-codes/famix/dist/FamixRepository';
import fs from 'fs-extra';
import { Container, injectable } from 'inversify';
import 'reflect-metadata';
import { Logger } from 'winston';

import Import from '../../../app/Console/import';
import { RedTwig } from '../../../app/Services/RedTwig';

@injectable()
class TestImportCommand extends Import {
    public argv: string[];
    public static args = Import.args;
}

let command: Import;
let container: Container;

const importer = {
    import: jest.fn((message: string) => void 0),
};

const logger = {
    info: jest.fn((message: string) => void 0),
};

const repository = {
    toJSON: jest.fn((message: string) => '[]'),
};

jest.mock('fs-extra', () => ({
    outputFile: jest.fn(async (...args) => void 0),
}));

describe('import command', () => {
    beforeAll(() => {
        container = new Container();

        container.bind<RedTwig>('RedTwig').toConstantValue(importer as any);
        container.bind<FamixRepository>('FamixRepository').toConstantValue(repository as any);

        container.bind<Logger>('Logger').toConstantValue(logger as any);
        container.bind('TestImportCommand').to(TestImportCommand);

        command = container.get<Import>('TestImportCommand');
    });

    afterAll(() => {
        jest.restoreAllMocks();
    });

    it('should not run update command without args', async () => {
        await expect(command.run).rejects.toThrowErrorMatchingSnapshot();
    });

    it('should not run update command without required args', async () => {
        command.argv = ['project'];
        await expect(command.run).rejects.toThrowErrorMatchingSnapshot();
    });

    it('should not run update command with invalid flags', async () => {
        command.argv = ['project', '--invalid'];
        await expect(command.run).rejects.toThrowErrorMatchingSnapshot();
    });

    it('should run command to import project', async () => {
        command.argv = ['path/to/project', '--output', 'path/to/output', '--format', 'json'];
        await command.run();

        expect(fs.outputFile).toHaveBeenCalledWith('path/to/output/model.json', '[]', { encoding: 'utf8' });
        expect(logger.info).toHaveBeenCalledWith(`Generated the model output in json format to path/to/output`);
    });
});
