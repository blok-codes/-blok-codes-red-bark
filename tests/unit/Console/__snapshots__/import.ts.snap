// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`import command should not run update command with invalid flags 1`] = `
"Unexpected argument: --invalid
See more help with --help"
`;

exports[`import command should not run update command without args 1`] = `
"Missing required flag:
 -o, --output OUTPUT  [2mPath to generate the model output file[22m
See more help with --help"
`;

exports[`import command should not run update command without required args 1`] = `
"Missing required flag:
 -o, --output OUTPUT  [2mPath to generate the model output file[22m
See more help with --help"
`;
